samples: mesh: boards: nrf52: upgrade to pass PTS, improved vendor model, randomize pulishers TID on boot

vikrant8052 opened the pull request:
Updated state Binding as per PTS test requirements. Now if gen_onoff_server &
gen_onpowerup_server received Prohibited values then they will not react on it further. Plus make necessary changes wherever required for message handlers as per Mesh Model
Specification which is mostly regarding to default & range values. Now right Status code would get publish in response to set lightness & temperature range. Also upgrade Message handler
gen_delta_set_unack() algorithm as per PTS requirements.

Update Vendor model message handlers & make it to work as per TID like other Models defined by SIG. Removed union based data extraction mechanism. Improved code readability by declaring some pre-processor definitions in device_composition.c

Remove model_instance variable from some structures define in device_composition.h & its relevant code. Now on reboot, NODE as client do not start message publishing with TID = 1. Instead of that it would start with any random value from 0 to 255. Added extra case under state binding's switch statement for Lightness i.e DELTA_LEVEL. Also upgrade binding between

1. root element's LEVEL state & Light Lightness Actual state.
2. Light Lightness Linear state & Light Lightness Actual state.
3. Light CTL lightness state & Light Lightness Actual state

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8051 commented:
@jhedberg @carlescufi 

Please review & merge this PR. Previous, PR:8781 failed due to unknown error from shippable.

For details:
https://app.shippable.com/github/zephyrproject-rtos/zephyr/runs/18527/summary/

vikrant8051 commented:
@jhedberg 
Please apply my PR first. And after that if @MariuszSkamra feels it not correct then he is most welcome to merge to his PR. I'm not against it.

vikrant8051 commented:
No I've not take his code for any reference.

vikrant8051 commented:
@jhedberg 
He has changed architecture of gen_onoff_server by adding extra,

static int gen_onoff_set_exec(struct bt_mesh_model *model,
			      struct bt_mesh_msg_ctx *ctx,
			      struct net_buf_simple *buf)

function. 

But in my case, I implemented it differently.

vikrant8051 commented:
In case of range upgrade for lightness & temperature , he has not implement algorithm to update status code. But I've implemented it.

vikrant8051 commented:
@jhedberg 
In case of range upgrade for lightness & temperature , he has not implement algorithm to update status code. But I've implemented it.

vikrant8051 commented:
@jhedberg 

Conflict arose because he didn't notify me that he is going to work on same App to solve the bugs. If he had, I would not do anything further till his PR get merge. He suddenly alert me when I had already invest my energy.

This doesn't means his contribution is less than me. And if you merge his PR first, then I don't have any problem.

MariuszSkamra commented:
@jhedberg @vikrant8051
I have a feeling this leads to nowhere.
I’ll abandon my PR. One last word from my side.
Please @vikrant8051 continue your work. But I have a feeling you should read a bit about open source and how to collaborate. Your attitude is rude, because you are not asking for help or review. You demand testing and merging. And this „private” message to our maintainer to not not merge my patch... Nasty. I felt offended. 

Regarding yesterdays situation I ping you on e-mail that I’m working on it when I saw your e-mail. Is something wrong in sending patches to „your” application? And no I did not changed the architecture. The patch was really simple.

I’ll review this patch later, because I’m in train now.

vikrant8051 commented:
@MariuszSkamra 
Let @jhedberg decide what to do. I don't have a problem if he merged your PR first. I'm ready to make changes as per that in my PR.

I need your thorough review. Thanks for it !!

I'm just contributing for fun & not for name. I have to enter name & email to satisfy Zephyr's norm. Otherwise not at all interested in it. My company too not going to use Zephyr in actual products (due to other reasons). Even after that if you wanna criticize me you are welcome.

vikrant8051 commented:
wait & I'm looking possibility of moving range comparison for Lightness in different handlers to state_binding() function.

vikrant8051 commented:
Mesh Model Specification is silent about binding following values with lightness_min_range & lightness_man_range for

1. light_lightness_default value
2. light_ctl_lightness value
3. light_ctl_default_lightness value

So removed conditions under respective handlers. Other things I've maintain as it is.

vikrant8051 commented:
@MariuszSkamra 
With due respect, I've added your contribution details in latest commit. Kindly consider it.

I think from my side done with bugs based on logs you've already shared.

I will notify you if going to change existing architecture after merging this PR so that we could avoid what happened previously. And I once again apologize for it. :)

MariuszSkamra commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -874,14 +937,30 @@ static void light_lightness_range_set_unack(struct bt_mesh_model *model,
+
+	if (min == 0) {
+		/* The provided value for Range Min cannot be set */
+		state->status_code = 0x01;
```
consider using a simple define for this, e.g 
#define STATE_STATUS_SUCCESS 0x00
#define STATE_STATUS_FAILED 0x01

vikrant8051 replied:
There are 3 values as of now, as per 7.2..... so I will define

#define STATE_STATUS_SUCCESS 0x00
#define CANNOT_SET _RANGE_MIN 0x01
#define CANNOT_SET _RANGE_MAX 0x02

MariuszSkamra commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -309,11 +341,11 @@ static void gen_onoff_set_unack(struct bt_mesh_model *model,
 	struct generic_onoff_state *state = model->user_data;
 
-	tmp8 = net_buf_simple_pull_u8(buf);
+	state->tmp = net_buf_simple_pull_u8(buf);
```
try to use more descriptive names or at least provide some comment what is the first byte in this message, so that if someone is reading your code knows what you assign to state->tmp without looking into the specification

vikrant8051 replied:
Here tmp is temporary variable only.
If it is greater than 1 then it will not allow to assign any value to actual onoff state variable
It also get access in gen_onoff_set(), based on which it will not publish if received gen_onoff
state is > 1. Hence put it in user's **struct generic_onoff_state**

jhedberg replied:
Please rename it to something more meaningful. `tmp` is really only ok for a temporary stack variable which the function uses for multiple different purposes. In this case it seems single-purpose and it's not temporary (it's part of of state struct instead of being a stack variable). It seems to me like it contains the current OnOff state value, so why not call it `val`?

vikrant8051 replied:
Ok. I will do it shortly.

MariuszSkamra commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.h:
```diff
@@ -23,13 +24,16 @@ struct generic_onoff_state {
 
 struct generic_onpowerup_state {
 	u8_t onpowerup;
+	u8_t tmp;
```
ditto

vikrant8051 replied:
Yes. 
Kindly consider same explanation as above.

Since have to access it in **gen_onoff_set_unack()** as well as in **gen_onoff_set()**

MariuszSkamra commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.h:
```diff
@@ -15,6 +15,7 @@ enum temperature { ONOFF_TEMP = 0x01, LEVEL_TEMP, CTL_TEMP, IGNORE_TEMP};
 
 struct generic_onoff_state {
 	u8_t onoff;
+	u8_t tmp;
```
Is it necessary to put temporary variable into this structure?

vikrant8051 replied:
Kindly consider same explanation as above.

MariuszSkamra reviewed (commented):
Take a look at comments I added. Apart from that I'm fine with this patch if fixing the issues we found.
 
Please remove my name from the commit message ^^

vikrant8051 commented:
@MariuszSkamra I've make suggested changes & all checks have passed.

vikrant8051 commented:
@jhedberg @MariuszSkamra I've make suggested changes & all checks have passed.

vikrant8051 commented:
Please wait trying to simplify Generic Delta set message handler.

vikrant8051 commented:
I've make simplified implementation of Generic Delta set message handler.

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.h:
```diff
@@ -10,26 +10,34 @@
 struct generic_onoff_state {
 	u8_t onoff;
-	u8_t previous;
+	u8_t current;
```
I see you already have `onoff` in this struct? Isn't that the current state? At least the "get" message handler uses it for the response. What exactly is `current` then if it's not the current value?

jhedberg replied:
If I understand it right (looking more closely at the code than the last time - I wish you would have corrected me then!), the only reason you're not using a local variable is that you need a way to find out the results of calling `gen_onoff_set_unack()` from `gen_onoff_set()`. One solution could be to consider refactoring these functions so that you have a third function that is essentially the same as the `unack()` except that it's able to return an error in case of invalid parameters. You can then use the error in the non-unack case to decide whether to send the status or not.

vikrant8051 replied:
I will make it val

jhedberg replied:
That does not solve the issue I pointed out. As far as I see you don't need this value for anything else than the `set_unack()` -> `set()` call path, so having it in the state struct is overkill (and confusing since it gives the impression that it's needed in some wider context). Please correct me if I'm wrong with these observations.

jhedberg commented:
Unrelated to this PR (i.e. please make a separate commit for it!) could you remove the union in `vnd_set_unack()`. That's completely unnecessary and something I noticed when reviewing the app code. You're getting the value using `pull_le16()` so you already have it in host-endianness, i.e. don't need to do any ugly tricks with a union.

vikrant8051 commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -684,10 +726,11 @@ static void light_lightness_set_unack(struct bt_mesh_model *model,
 				      struct net_buf_simple *buf)
 {
 	u8_t tid;
+	u16_t tmp16;
```
I hope it will be final from you & @jhedberg  :)
Because I've to sleep now. :) :)

vikrant8051 replied:
done !!

vikrant8051 commented:
@MariuszSkamra I've make changes wherever tmp16 was there.
Hope now Shippable will pass it & you too. :)

vikrant8051 commented:
@jhedberg Please approved it now if it is OK for you. I fixed Vendor model issue too via same PR.

jhedberg commented:
@vikrant8051 it seems you've missed a key comment from me, regarding the necessity of another state struct variable. Let me quote my two comments here in case you have difficulty finding them in the comment history:

@jhedberg wrote:
> If I understand it right (looking more closely at the code than the last time - I wish you would have corrected me then!), the only reason you're not using a local variable is that you need a way to find out the results of calling gen_onoff_set_unack() from gen_onoff_set(). One solution could be to consider refactoring these functions so that you have a third function that is essentially the same as the unack() except that it's able to return an error in case of invalid parameters. You can then use the error in the non-unack case to decide whether to send the status or not.

@vikrant8051 wrote:
> I will make it val

@jhedberg wrote:
> That does not solve the issue I pointed out. As far as I see you don't need this value for anything else than the set_unack() -> set() call path, so having it in the state struct is overkill (and confusing since it gives the impression that it's needed in some wider context). Please correct me if I'm wrong with these observations.

jhedberg commented:
@vikrant8051 also, I expressly asked you to make the vendor model fix a separate commit, yet you've baked it into the existing single commit. Please split it back out.

vikrant8051 commented:
@jhedberg 
It's user structure & app dependent. Others will understand it. Calling another function will be overhead  to just Turn On & Off. Plus it looks nice.

jhedberg reviewed (changes requested):
I'll wait until the issues I've brought up have been resolved

vikrant8051 commented:
Ok.. then I will make it.

vikrant8051 commented:
Is it Ok if I define temporary static globle variable in device_compositon.c ?

vikrant8051 commented:
@jhedberg Is it Ok if I define temporary static globle variable in device_compositon.c ?

jhedberg commented:
> @jhedberg Is it Ok if I define temporary static globle variable in device_compositon.c ?
For e.g. get_send_enable.

what's get_send_enable? What will the variable be used for? is this some new thing, or relating to my earlier comments?

Regarding the state struct, it should really only contain state information that persists between messages. Putting a variable there that's only relevant/valid for a single call path confused me, and will likely confuse others as well. Such variables should be on the stack instead of in a state struct. The overhead of making a function call from `set()` will not be any more overhead than what you have now from `set_unack()`. What would be nicer if the message handlers had the ability to directly return an error instead of returning void, however we can't really make that change as it would make any existing users of the API fail to build (or at least throw a compiler warning) due to a mismatch in function pointer type.

vikrant8051 commented:
I've updated my previous comment with code. Please consider that.

vikrant8051 commented:
@jhedberg I've updated my previous comment with code. Please consider that.

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -587,8 +628,13 @@ static void gen_onpowerup_set(struct bt_mesh_model *model,
-	gen_onpowerup_get(model, ctx, buf);
+
+	if (state->val <= 0x02) {
+		gen_onpowerup_get(model, ctx, buf);
```
You're still using a state struct variable instead of a stack variable here. Also, isn't this broken since `set_unack()` will change the parsing state of the net_buf (with its `net_buf_pull_*()` calls), so what you're giving to `gen_powerup_get()` will not be what it expects (in fact it'll likely result in reading invalid memory). Did you actually test this?

vikrant8051 replied:
Done !!

jhedberg commented:
Also, I'm still waiting for the vendor model changes to be split to a different commit (it can still be this same PR however).

vikrant8051 commented:
@jhedberg Is following way is Ok for you ? In this case, I could remove var from structure.


```
static u8_t tmp_variable;
.
.
.
.
static void gen_onoff_set_unack(struct bt_mesh_model *model,
				struct bt_mesh_msg_ctx *ctx,
				struct net_buf_simple *buf)
{
	u8_t tid;
	struct net_buf_simple *msg = model->pub->msg;
	struct generic_onoff_state *state = model->user_data;

	tmp_variable = net_buf_simple_pull_u8(buf);
	tid = net_buf_simple_pull_u8(buf);

	if (state->last_tid == tid && state->last_tx_addr == ctx->addr) {
		return;
	}

	state->last_tid = tid;
	state->last_tx_addr = ctx->addr;

	if (tmp_variable > 0x01) {
		return;
	}
	state->onoff = tmp_variable;

	if (state->model_instance == 0x01) {
		/* Root element */
		state_binding(ONOFF, ONOFF_TEMP);
		update_light_state();
	} else if (state->model_instance == 0x02) {
		/* Secondary element */
	}

	if (model->pub->addr != BT_MESH_ADDR_UNASSIGNED) {
		int err;

		bt_mesh_model_msg_init(msg,
				       BT_MESH_MODEL_OP_GEN_ONOFF_STATUS);
		net_buf_simple_add_u8(msg, state->onoff);

		err = bt_mesh_model_publish(model);
		if (err) {
			printk("bt_mesh_model_publish err %d\n", err);
		}
	}
}

static void gen_onoff_set(struct bt_mesh_model *model,
			  struct bt_mesh_msg_ctx *ctx,
			  struct net_buf_simple *buf)
{
	struct generic_onoff_state *state = model->user_data;

	gen_onoff_set_unack(model, ctx, buf);

	if (tmp_variable <= 0x01) {
		gen_onoff_get(model, ctx, buf);
	}
}
```

vikrant8051 commented:
@jhedberg 
I've make suggested changes. Please review.

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -554,18 +637,18 @@ static void gen_onpowerup_set_unack(struct bt_mesh_model *model,
 				    struct bt_mesh_msg_ctx *ctx,
 				    struct net_buf_simple *buf)
 {
```
Wouldn't it make sense to have this function call your new `onpowerup_set()` function, so you don't need to duplicate the code? That was the refactoring that I was suggesting earlier.

vikrant8051 replied:
Done !!

jhedberg commented on samples/boards/nrf52/mesh/onoff_level_lighting_vnd_app/src/mesh/device_composition.c:
```diff
@@ -583,12 +666,48 @@ static void gen_onpowerup_set_unack(struct bt_mesh_model *model,
-	gen_onpowerup_set_unack(model, ctx, buf);
-	gen_onpowerup_get(model, ctx, buf);
+	if (onpowerup_set(model, ctx, buf) == 0) {
+		gen_onpowerup_get(model, ctx, buf);
```
The `onpowerup_set()` function will have changed the parsing state of `buf` so calling `gen_onpowerup_get()` here will cause reading past the end of the buffer. This is the second or third time I'm telling you this, please don't make me do it again :)

Your options are to either pre-parse the value and pass a `u8_t` instead of `net_buf_simple` to `onpowerup_set()`, or then (possibly better since you don't need to change `gen_onpowerup_get()`) use `net_buf_simple_save()` and `net_buf_simple_restore()` like I suggested earlier.

vikrant8051 replied:
working on it.

vikrant8051 replied:
I think what had done earlier is better than all this. 
Just to explain developers, why to do more complications ?

What going to happen if I add simply one extra tmp variable in struct for onoff state or onpowerup state ? It is by the way user defined data. 

I'm going to make things as it was. Kindly consider it.

vikrant8051 replied:
@jhedberg 

 I think what I had done earlier is better than all this. 
Just to explain developers, why to do more complications ?

What going to happen if I add simply one extra tmp variable in struct for onoff state or onpowerup state ? It is by the way user defined data. 

I'm going to make things as it was. Kindly consider it.

vikrant8051 replied:
@jhedberg 
I will make changes as per your suggestion in NEXT PR.  Kindly consider as of now.

vikrant8051 replied:
Done !!

vikrant8052 committed:
samples: mesh: boards: nrf52: upgrade to pass PTS test

Updated state Binding as per PTS test requirements. Now if
gen_onoff_server & gen_onpowerup_server received Prohibited values
then they will not react on it further. Plus make necessary changes
wherever required for message handlers as per Mesh Model
Specification which is mostly regarding to default & range values.
Now right Status code would get publish in response to set lightness
& temperature range. Also upgrade Message handler
gen_delta_set_unack() algorithm as per PTS requirements.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8052 committed:
samples: mesh: boards: nrf52: improved vendor model

Update Vendor model message handlers & make it to work as per
TID like other Models defined by SIG. Removed union based data
extraction mechanism.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8051 commented:
@jhedberg
For Vendor model, I've add another commit. For others first one is enough since it was purely based on as per PTS requirements.

I know it was not as per your expectations but believe me it will be easy to digest for newbies like me.
And of course, it is far better than what already there in master branch.
 
Features wise I've added enough Models & don't gonna add anything further.
For resolving Bugs, I will be there forever.

Hope it will pass in PTS test too (except for unimplemented state transition & gen_move_set_unack message handler). Now really need a break since from past few weeks working on same thing.

@carlescufi @MariuszSkamra 

Thank You to all !!

vikrant8051 commented:
@jhedberg 
How then you allow 

**struct onoff_state {
	u8_t current;
	u8_t previous;   // <--- Why this could not be confusing to you ?
	u8_t led_gpio_pin; // <----- same here ?
	struct device *led_device;
};**

this structure in samples/boards/nrf52/mesh/onoff-app ?

That means PTS going to fail for this app too.

vikrant8051 commented:
@jhedberg I really don't want to edit my current implementation but just because of your requirement I'm not able concentrate on my other work since it is continuously ringing something in my head.

jhedberg commented:
@vikrant8051 regarding my previous comments on the parsing state, I was looking through the code in more detail now (comparing with onoff-app which you've used as reference), and since the "get" functions don't need to parse anything from the net_buf_simple (in fact they do nothing with it) there's no issue with parsing, so just ignore my earlier comment.

jhedberg commented:
> this structure in samples/boards/nrf52/mesh/onoff-app ?
> Here Author has added whatever he want. Why ?

You give us (maintainers) too much credit for catching anything/everything in all  PRs. The onoff-app is e.g. lacking Prohibited value checking, so I suspect it will fail some PTS tests as well (@MariuszSkamra you might wanna run the tests against that app too). Since the app is not doing any Prohibited value checking it's simply tracking the old and new values of the state to detect changes. This is different from what you want to do. You want to detect when there was no change due to a Prohibited value in the paramers (i.e. the message should be ignored). Either way, tracking message parsing information in the context struct makes little sense and is just confusing.

vikrant8051 commented:
@jhedberg 
So may I conclude that what currently implemented is Okay for you and
your are now waiting for only PTS result ?

jhedberg commented:
> So may I conclude that what currently implemented is Okay for you

@vikrant8051 no, my earlier comments regarding refactoring still hold, i.e. I do think your current code will cause confusion, no matter how clear it's for you. While the onoff-app has its issues (which should be fixed), it's not doing what you're wanting to do, i.e. trying to pass the failure/success result of a function back to the calling function by using a global state struct that's intended to reflect the current state of the model.

vikrant8051 commented:
@jhedberg 
`I do think your current code will cause confusion, no matter how clear it's for you.
`
Is it bug ? I don't think so. Could you elaborate about your confusion ?
So that I could add comments in code to clear doubts of developers. 


Could you please edit onoff-app for me to show what refactoring you are expecting from me ?

vikrant8051 commented:
@jhedberg 
I've done refactoring in local git directory. Please check your email. Once you approved then I will push changes to PR.

MariuszSkamra commented:
@jhedberg I'll run onoff-app against PTS today.

vikrant8051 commented:
On Tue, Jul 10, 2018, Vikrant More wrote:
> If I'm going to make following changes. I think this is what you are
> expecting from me. Am I right ?

Yes, that's correct.

A couple of other points:

> static int *onoff_set_unack*(struct bt_mesh_model *model,
>                    struct bt_mesh_msg_ctx *ctx,
>                    struct net_buf_simple *buf)
> {
>     u8_t tid, onoff;
>     struct net_buf_simple *msg = model->pub->msg;
>     struct generic_onoff_state *state = model->user_data;
> 
>     onoff = net_buf_simple_pull_u8(buf);
>     tid = net_buf_simple_pull_u8(buf);
> 
>     if (state->last_tid == tid && state->last_tx_addr == ctx->addr) {
>         return 0;
>     }
> 
>     state->last_tid = tid;
>     state->last_tx_addr = ctx->addr;
> 
>     if (onoff > 0x01) {
>         return -1;
>     }

I think you have the checks the wrong way around. If you get a
prohibited value for onoff, then this function should always return
failure (i.e. no response). Currently it would return success if onoff
is Prohibited but last_tid == tid || last_tx_addr == ctx->tx_addr.

Furthermore, don't use -1 as an error. Most code assumes that negative
integer errors map to a POSIX error number. So if you want to use that
then use -EINVAL. The other option is to return bool and use false for
failure and true for success.

Another thing: it might be good to add defines for 0x00 and 0x01. That
would help improve the readability of the code. E.g. for checking for
valid values you'd have:

        if (onoff == STATE_ON || onoff == STATE_OFF) {
                ... valid handling ...
        }

And for checking invalid values:

        if (onoff != STATE_ON && onoff != STATE_OFF) {
                return -EINVAL;
        }

One other thing:

>     if (state->model_instance == 0x01) {

I didn't notice this instance thing previously. What's the magic 0x01
value? Normally the root element is element 0 and not 1. Wouldn't it
make more sense to compare the element pointer to your element array, or
if it's in a different c-file have a helper function to convert the
pointer into an index?

vikrant8051 commented:
@jhedberg 
Please now don't start with other things. Keep something for developers also who are going to refer this app. Please !! 

**I've avoid your these points :** 
Another thing: it might be good to add defines for 0x00 and 0x01. That
would help improve the readability of the code. E.g. for checking for
valid values you'd have:

    if (onoff == STATE_ON || onoff == STATE_OFF) {
            ... valid handling ...
    }

And for checking invalid values:

    if (onoff != STATE_ON && onoff != STATE_OFF) {
            return -EINVAL;
    }

One other thing:

    if (state->model_instance == 0x01) {

I didn't notice this instance thing previously. What's the magic 0x01
value? Normally the root element is element 0 and not 1. Wouldn't it
make more sense to compare the element pointer to your element array, or
if it's in a different c-file have a helper function to convert the
pointer into an index?

jhedberg reviewed (approved):
There are still lots of things to clean up & improve, but for the purposes of the issues raised in this PR I think it's now good enough for merging.

vikrant8051 commented:
@jhedberg 
>> There are still lots of things to clean up & improve, 

Are there any bugs OR are they just architectural issues ?

Please share those things in one shot and not in chunks.  So that I can improve all in next PR. Sorry for bothering you again & again by creating new PR. But I don't have any other option since for reference there is no any App available for advanced models anywhere in world or on internet.  Plus Bluetooth Mesh model specification is not so clear for programmers since all relevant things are not available at single location.

jhedberg commented:
> Are there any bugs OR are they just architectural issues ?

There are no bugs I'm aware of, but that's anyway something we'll see once the PTS tests are run again.

The rest is cosmetic, and most things I've already pointed out to you. I can probably fix up any left-overs myself in some later PR.

vikrant8051 commented:
@jhedberg That's sounds great. Thank You !!

vikrant8051 commented:
@jhedberg 

>> The rest is cosmetic, and most things I've already pointed out to you.

I've edited CODE as per your latest 2 requirements. In case of Model instance, I've define ..

#define ROOT_ELEMENT_INSTANCE		0x00
#define SECONDARY_ELEMENT_INSTANCE	0x01

in device_composition.h file.
 
May I submit a new commit or let it be as it is?

jhedberg commented:
You can either add another commit to this PR or wait until this is merged and then send the commit as a separate PR. Regarding the instance stuff, I'm thinking maybe it would make sense to introduce a helper API to the Mesh stack called `bool bt_mesh_elem_is_primary(const struct bt_mesh_elem *elem)`.

vikrant8051 commented:
if (bt_mesh_model_elem(model)->addr == elements[0].addr) {
		/* Root element */
		state_binding(ONOFF, ONOFF_TEMP);
		update_light_state();
	} else if (bt_mesh_model_elem(model)->addr == elements[1].addr) {
		/* Secondary element */
		printk("Hello World\n");
	}

vikrant8051 commented:
I will merge with this only

vikrant8052 committed:
samples: mesh: improved code readability & remove redudancy

Improved code readability by declaring some pre-processor
definitions in device_composition.c

Remove model_instance variable from some structures define in
device_composition.h & its relevant code.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8051 commented:
When **board/remote/client** get power &  send something to Server on other boards it always put TID = 1.

If after that **client** get switch off & on, and try to send something again to same servers then those server do not reacts since this time too TID == 1.

Is this bug ?

vikrant8051 commented:
@jhedberg 
When board/remote/client get power & send something to Server on other boards it always put TID = 1.

If after that client get switch off & on, and try to send something again to same servers then those server do not reacts since this time too TID == 1.

Is this bug  or wrong implementation of client side ?

Solution: instead of sending 1 after every reboot, send random no between {0..255}

vikrant8051 commented:
@jhedberg 
I found solution for previous issue .....

For that, I've to add following function in publisher.c
```

void randomize_publisher_TIDs(void)
{
	bt_rand(&tid_level, sizeof(tid_level));

#ifndef VND_MODEL_TEST
	bt_rand(&tid_onoff, sizeof(tid_onoff));
#else
	bt_rand(&tid_vnd, sizeof(tid_vnd));
#endif
}

----------------------------------------------------------------------------------------------------------------------
& main() in main.c as ......

void main(void)
{
	int err;

	light_default_status_init();

	gpio_init();

	update_light_state();

	printk("Initializing...\n");

	/* Initialize the Bluetooth Subsystem */
	err = bt_enable(bt_ready);
	if (err) {
		printk("Bluetooth init failed (err %d)\n", err);
	}

	**randomize_publisher_TIDs();**
}
```

vikrant8052 committed:
samples: mesh: boards: nrf52: randomize publishers TID on boot

Now on reboot, NODE as client do not start message publishing
with TID = 1. Instead of that it would start with any random value
from 0 to 255.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8052 committed:
samples: mesh: boards: nrf52: modifications as per 3.3.2.2.3

Added extra case under state binding's switch statement for
Lightness i.e DELTA_LEVEL. Also upgrade binding between

1. root element's LEVEL state & Light Lightness Actual state.
2. Light Lightness Linear state & Light Lightness Actual state.
3. Light CTL lightness state & Light Lightness Actual state

This is as per Bluetooth Mesh Model Specification 3.3.2.2.3.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

vikrant8051 commented:
@jhedberg @MariuszSkamra 
Please not that current PR:8793 received green marked based on first 2 commits. So kindly review last 3 commits. Your feedback is important.

Is Mariusz going to execute PTS aginst this latest PR along with onoff-app? 

Hope it will get merge soon.

MariuszSkamra commented:
@vikrant8051 @jhedberg I've already tested the on-off app against Mesh Model PTS Test Cases, and the results are as we expected (fails). So if we want to have all the samples compliant with Specification  the on-off application should be fixed

vikrant8051 commented:
@MariuszSkamra Thanks for updating.

vikrant8051 commented:
wait ..trying to implement 3.3.2.2.3

vikrant8051 commented:
wait

vikrant8051 commented:
Kindly wait ..since reverse Binding between in following bound states is not yet clear.

1) Light Lightness Actual <-> Generic Level
2) Light Lightness Actual <-> Light Lightness Linear
3) Light Lightness Actual <-> Light CTL Lightness

I would like to confirm it & will make changes. 

For e.g.

```
Let
L= Level state
LLA = light lightness actual
LLA_min  = 1000
LLA_max  = 50000

As per specs, L & LLA are bound States 

L = LLA - 32768
LLA = L + 32768

------------------------------------------------------

What should be value of status responses,
when user set L= -32700 out of following options ?

A) L= -31768, LLA= 1000

B) L= -32700, LLA= 1000
```

MariuszSkamra commented:
@vikrant8051 is this PR ready to be tested?

vikrant8051 commented:
@MariuszSkamra Yes, sure !!

vikrant8052 committed:
samples: mesh: boards: nrf52: edit struct for gen. level

Edit data types defined in struct for gen. level.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

MariuszSkamra commented:
MMDL/SR-CL/COMP/BV-01-C PASS
MMDL/SR/GOO/BV-01-C PASS
MMDL/SR/GOO/BV-02-C FAIL
MMDL/SR/GOO/BI-01-C PASS
MMDL/SR/GLV/BV-01-C PASS
MMDL/SR/GLV/BV-02-C FAIL
MMDL/SR/GLV/BV-03-C FAIL https://hastebin.com/ohanupezot.sql https://hastebin.com/tubipuvequ.css
MMDL/SR/GLV/BV-04-C PASS
MMDL/SR/GLV/BV-05-C FAIL https://hastebin.com/lovuyurana.php https://hastebin.com/evawefahik.css

vikrant8051 commented:
@MariuszSkamra  Ok. I'm working on it.

vikrant8051 commented:
@MariuszSkamra Is all log relevant to "#onoff_level_lighting_vnd_app + this PR" ?
All are creating confusion in Mind. Please run PTS only after merging this PR.

vikrant8052 committed:
samples: mesh: boards: nrf52: avoid responding to wrong messages

If temperature value are out of range which is from 0x0320
to 0x4E20 then no relevant get message handler will execute.

Signed-off-by: Vikrant More <vikrant8051@gmail.com>

jhedberg commented:
> Please run PTS only after merging this PR.

@vikrant8051 the purpose of the PTS testing is to decide the maturity of this PR, i.e. if it's ready to be merged. So what you're requesting makes no sense.

vikrant8051 commented:
@jheadburg
Currently most of the failure are because of ...
>> Invalid state after reset (persistent storage not supported)
>> Invalid behaviour (sending status) when Invalid parameters are passed via Set command

I've send latest commit for second one. If possible please merge this one. I will take care of other issues in next PR. Current PR is already heavy that's why I'm requesting you " Please merge as of now"

vikrant8051 commented:
@jhedberg 
Most of the errors are because of absence of 
1) persistence storage 
2) transition time based implementation.
3) proper rounding off in case float values

For prohibited values I've already added latest commit.

Other issue are 
1) content of status response for range set messages when min & max are not valid
  - currently I'm sending three things even when status code is not equal to zero.
    -- status code
    -- previous min range value
    -- previous max range value

PR:8793 is well matured now. :) 
I'm in process of clearing my doubts by asking questions of Bluetooth forum via their support emails.
I will rectify things in upcoming days. So as of now please merge this PR.

MariuszSkamra commented:
@jhedberg I think, this PR looks OK. It fixes some of the issues found. Further fixes may be pushed in another PR. Now, when we have PTS results for this PR, it is easier for @vikrant8051 to send a patch related to one of the issues we found (it can be missing functionality, bugfix, etc.).
Doing this way I can then run tests for only e.g Light Lightness Model if the PR relates to this and only this model. Current workflow leads to testing everything after every change Vikrant made. I see now there's another patch added that is unrelated to the previous ones.
Testing takes some time (PTS is rather fast but I have to gather the logs too).
So to sum up: @vikrant8051 Please send patches that relates to one of the things you pick to fix it would be easier for me to test it (I would have less tests to run, and more time so that we could analyze the logs together).

MariuszSkamra commented:
> @MariuszSkamra Is all log relevant to "#onoff_level_lighting_vnd_app + this PR" ?
> Some are creating confusion in Mind. Please run PTS only after merging this PR.
No, those are all the Server side tests. TBH I am a bit lost what you are actually fixing in this PR, so I ran all the tests. PR title states for "upgrade to pass PTS" ;)

vikrant8051 commented:
@MariuszSkamra Thanks for your reply !! Hope @jhedberg will agree on that. 

I understand your concern. But to satisfy PTS all changes I've done looked valid (except for Vendor Model ). I'm still afraid to use git rebase otherwise I could do better.

vikrant8051 commented:
@MariuszSkamra 
Entire App is develop my keeping table 6.124 in front of Eyes. 
I have to add Clients since no App is available to control such advanced Server Models using built-in GUIs.

vikrant8051 commented:
@MariuszSkamra @jhedberg I'm going to restructure every things to simplify commits related things. 
Final code will be as it.

jhedberg commented:
@vikrant8051 @MariuszSkamra ok, merged now. Make a new PR for any follow-up fixes.

vikrant8051 commented:
@jhedberg @MariuszSkamra Thanks for merging this PR & helping me to remove bugs in it.
